onadata/apps/api/permissions.py

```diff
@@ -513,7 +515,11 @@ def has_permission(self, request, view):
             else:
                 # Raises a permission denied exception, forces authentication
                 return False
-            profile, _ = UserProfile.objects.get_or_create(user=user)
+            profile = cache_single_object(
+                USER_PROFILE_PREFIX,
+                user.id,
+                UserProfile.objects.get_or_create(user=user)
+                )
```

Comment:
This would possibly also cache a User who is already present in the database right ? Not entirely sure if we need to cache a user that's already retrievable from the db in this case.... It might end up in a scenario where we'll always be caching all users who make authenticated submissions.

_I think we should only cache here if we're creating a Profile_

https://github.com/onaio/onadata/pull/1823/files#diff-356ba073bddf0df5ba9d9c050982939eR47-R51
https://github.com/onaio/onadata/pull/1823/files#diff-cec4d5661d03adabc14e2f6efc04e7b6R177-R181
https://github.com/onaio/onadata/pull/1823/files#diff-cec4d5661d03adabc14e2f6efc04e7b6R240-R243

Same case for the above ^ _I guess an easy change would be to make the `cache_single_object` function capable of knowing whether it should cache or not ? Possibly checking if the query returns something_

Reply:
The idea was to avoid requests hitting the database for subsequent requests after a user profile is created. So, even for users who are saved in the db, in the first try to retrieve the UserProfile would hit the db, but subsequent requests wouldnt necessarily have to.

I cached the UserProfile objects here since at this point there are database requests to retrieve the UserProfile object for all these users, wouldn't we want to avoid these requests constantly hitting the db?

At the moment, the `cache_single_object` function checks if the user_profile cache for the user has been created, if not, then it makes the query. I was avoiding making requests to the db until absolutely necessary

Reply:
Okay.... but still this will make the request wont it ? `UserProfile.objects.get_or_create(user=user)` would make the db request as such it doesn't really decrease the amounts of request to the db....

Reply:
I understand now..cool cool, i'll restructure that abit